crates/locutus-router/src/isotonic_estimator.rs

```diff
@@ -0,0 +1,317 @@
+use locutus_core::ring::PeerKeyLocation;
+use locutus_core::Location;
+use pav_regression::pav::{IsotonicRegression, Point};
+use serde::Serialize;
+use std::collections::HashMap;
+
+const MIN_POINTS_FOR_REGRESSION: usize = 5;
+
+/// `IsotonicEstimator`  provides outcome estimation for a given action, such as
+/// retrieving the state of a contract, based on the distance between the peer
+/// and the contract. It uses an isotonic regression model from the `pav.rs`
+/// library to estimate the outcome based on the distance between the peer and
+/// the contract, but then also tracks an adjustment for each peer based on the
+/// outcome of the peer's previous requests.
+
+#[derive(Debug, Clone, Serialize)]
+pub(crate) struct IsotonicEstimator {
+    pub(crate) global_regression: IsotonicRegression,
+    pub(crate) peer_adjustments: HashMap<PeerKeyLocation, Adjustment>,
+}
+
+impl IsotonicEstimator {
+    // Define a constant for the adjustment prior size.
+    const ADJUSTMENT_PRIOR_SIZE: u64 = 10;
+
+    /// Creates a new `PeerOutcomeEstimator` from a list of historical events.
+    pub fn new<I>(history: I, estimator_type: EstimatorType) -> Self
+    where
+        I: IntoIterator<Item = IsotonicEvent>,
+    {
+        let mut all_points = Vec::new();
+
+        let mut peer_events: HashMap<PeerKeyLocation, Vec<IsotonicEvent>> = HashMap::new();
+
+        for event in history {
+            let point = Point::new(event.route_distance(), event.result);
+
+            all_points.push(point);
+            peer_events.entry(event.peer).or_default().push(event);
+        }
+
+        let global_regression = match estimator_type {
+            EstimatorType::Positive => IsotonicRegression::new_ascending(&all_points),
+            EstimatorType::Negative => IsotonicRegression::new_descending(&all_points),
+        };
+
+        let mut peer_adjustments: HashMap<PeerKeyLocation, Adjustment> = HashMap::new();
+
+        // Use the constant defined earlier.
+        let adjustment_prior_size = Self::ADJUSTMENT_PRIOR_SIZE;
+
+        // Use more descriptive variable names.
+        for (peer_location, events) in peer_events.iter() {
+            let mut event_count: u64 = adjustment_prior_size;
+            let mut total_adjustment: f64 = 0.0;
+            for event in events {
+                let peer_adjustment = event.result
+                    - global_regression
+                        .interpolate(event.route_distance())
+                        .unwrap();
```

Comment:
is this unwrap supposed to always work? the node would crash otherwise, in any case do `expect("node why we are panicking")`

Reply:
Good catch, I've added some checks.